File: LeetCode - Algorithm/Interval/56.py
# 56. Merge Intervals
# https://leetcode.com/problems/merge-intervals/

from typing import List


class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals = sorted(intervals, key=lambda x: x[0])

        result = [intervals[0]]

        i = 1
        j = 0

        while i < len(intervals):
            targetBeginning = intervals[i][0]
            resultEnd = result[j][1]

            if targetBeginning <= resultEnd:
                result[j][1] = max(result[j][1], intervals[i][1])
                i += 1
            else:
                result.append(intervals[i])
                i += 1
                j += 1

        return result


"""
[[1,3],[2,6],[8,10],[15,18]]
               ^
[[1,3]]
   ^
"""

File: LeetCode - Algorithm/Interval/56.js
// 56. Merge Intervals
// https://leetcode.com/problems/merge-intervals/

/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function (intervals) {
  const sortedIntervals = intervals.sort((a, b) => a[0] - b[0]);

  const result = [];
  result.push(sortedIntervals[0]);

  for (let i = 1; i < sortedIntervals.length; i++) {
    const [_, firstEnd] = result[result.length - 1];
    const [secondStart, secondEnd] = sortedIntervals[i];

    if (secondStart > firstEnd) result.push(sortedIntervals[i]);
    else result[result.length - 1][1] = Math.max(firstEnd, secondEnd);
  }

  return result;
};
